### PR TITLE
fix(messaging, analytics): remove cyclical imports

### LINKED ISSUE #6897: [🐛] Messaging - Require Cycle
## Issue

Describe your issue here

Getting require cycle warning on @react-native-firebase/messaging

```
 WARN  Require cycle: node_modules/@react-native-firebase/messaging/lib/index.js -> node_modules/@react-native-firebase/messaging/modular/index.js -> node_modules/@react-native-firebase/messaging/lib/index.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
```

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
{
  "name": "AppName",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "start": "react-native start",
    "test": "jest",
    "lint": "eslint .",
    "postinstall": "patch-package",
    "android:clean": "cd android && ./gradlew clean && cd ..",
    "android:apk": "cd android && ./gradlew assembleRelease && cd ..",
    "android:bundle": "cd android && ./gradlew bundleRelease && cd ..",
    "ios:install": "cd ios && pod install && cd .."
  },
  "dependencies": {
    "@apollo/client": "^3.6.9",
    "@eva-design/eva": "^2.1.1",
    "@gorhom/bottom-sheet": "^4",
    "@notifee/react-native": "^7.3.0",
    "@openspacelabs/react-native-zoomable-view": "^2.1.1",
    "@react-native-clipboard/clipboard": "^1.11.1",
    "@react-native-community/datetimepicker": "^6.2.0",
    "@react-native-community/netinfo": "^9.3.7",
    "@react-native-firebase/analytics": "^17.0.0",
    "@react-native-firebase/app": "^17.0.0",
    "@react-native-firebase/crashlytics": "^17.0.0",
    "@react-native-firebase/messaging": "^17.0.0",
    "@react-native-masked-view/masked-view": "^0.2.6",
    "@react-native-picker/picker": "^2.4.1",
    "@react-navigation/bottom-tabs": "^6.3.1",
    "@react-navigation/devtools": "^6.0.11",
    "@react-navigation/native": "^6.0.10",
    "@react-navigation/native-stack": "^6.6.2",
    "@shopify/flash-list": "^1.4.0",
    "@tanstack/react-query": "^4.16.1",
    "@ui-kitten/components": "^5.1.2",
    "apollo3-cache-persist": "^0.14.1",
    "axios": "^0.27.2",
    "dayjs": "^1.11.5",
    "graphql": "^16.5.0",
    "libphonenumber-js": "^1.10.18",
    "lodash": "^4.17.21",
    "lottie-ios": "3.4.0",
    "lottie-react-native": "^5.1.4",
    "native-base": "3.4.22",
    "payment": "^2.4.6",
    "react": "17.0.2",
    "react-hook-form": "^7.36.0",
    "react-native": "0.68.6",
    "react-native-collapsible": "^1.6.0",
    "react-native-config": "1.5.0",
    "react-native-confirmation-code-field": "^7.3.0",
    "react-native-device-info": "^10.3.0",
    "react-native-fast-image": "^8.6.1",
    "react-native-flash-message": "^0.2.1",
    "react-native-gesture-handler": "^2.4.2",
    "react-native-image-gallery": "^2.1.5",
    "react-native-inappbrowser-reborn": "^3.7.0",
    "react-native-linear-gradient": "^2.6.2",
    "react-native-lottie-splash-screen": "^1.0.1",
    "react-native-mmkv": "2.5.1",
    "react-native-mmkv-flipper-plugin": "^1.0.0",
    "react-native-modalize": "^2.0.13",
    "react-native-onboarding-swiper": "^1.2.0",
    "react-native-phone-number-input": "^2.1.0",
    "react-native-reanimated": "2.6.0",
    "react-native-reanimated-carousel": "^2.5.0",
    "react-native-render-html": "^6.3.4",
    "react-native-responsive-dimensions": "^3.1.1",
    "react-native-safe-area-context": "^4.2.5",
    "react-native-scalable-image": "^1.1.0",
    "react-native-screens": "3.14.1",
    "react-native-share": "^8.0.0",
    "react-native-super-grid": "^4.4.0",
    "react-native-svg": "13.7.0",
    "react-native-swipe-list-view": "^3.2.9",
    "react-native-ui-lib": "^6.17.1",
    "react-native-vector-icons": "^9.1.0",
    "react-phone-number-input": "^3.2.3",
    "traverse": "^0.6.6",
    "zustand": "4.1.2"
  },
  "devDependencies": {
    "@babel/core": "^7.12.9",
    "@babel/runtime": "^7.12.5",
    "@react-native-community/eslint-config": "^2.0.0",
    "babel-jest": "^26.6.3",
    "babel-plugin-module-resolver": "^4.1.0",
    "deprecated-react-native-prop-types": "^2.3.0",
    "eslint": "^7.32.0",
    "jest": "^26.6.3",
    "metro-react-native-babel-preset": "^0.67.0",
    "patch-package": "^6.5.0",
    "react-native-flipper": "0.162.0",
    "react-native-flipper-zustand": "^2.0.1",
    "react-native-svg-transformer": "^1.0.0",
    "react-test-renderer": "17.0.2"
  },
  "jest": {
    "preset": "react-native"
  }
}

```

#### `firebase.json` for react-native-firebase v6:

```json
{
    "react-native": {
      "crashlytics_debug_enabled": true,
      "crashlytics_disable_auto_disabler": true,
      "crashlytics_auto_collection_enabled": true,
      "crashlytics_is_error_generation_on_js_crash_enabled": true,
      "crashlytics_javascript_exception_handler_chaining_enabled": true
    }
}
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

platform :ios, '11.0'
install! 'cocoapods', :deterministic_uuids => false

target 'AppName' do
  config = use_native_modules!

  # Flags change depending on the env values.
  flags = get_default_flags()

  pod 'Firebase', :modular_headers => true
  pod 'FirebaseCore', :modular_headers => true
  pod 'GoogleUtilities', :modular_headers => true
  $RNFirebaseAsStaticFramework = true

  use_react_native!(
    :path => config[:reactNativePath],
    # to enable hermes on iOS, change `false` to `true` and then install pods
    # :hermes_enabled => flags[:hermes_enabled],
    :hermes_enabled => true,
    :fabric_enabled => flags[:fabric_enabled],
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )





  target 'AppNameTests' do
    inherit! :complete
    # Pods for testing
  end

  # Enables Flipper.
  #
  # Note that if you have use_frameworks! enabled, Flipper will not work and
  # you should disable the next line.
  use_flipper!()

  post_install do |installer|
    react_native_post_install(installer)
    __apply_Xcode_12_5_M1_post_install_workaround(installer)
    # Add these lines for Xcode 14 builds
    installer.generated_projects.each do |project|
      project.targets.each do |target|
          target.build_configurations.each do |config|
              config.build_settings["DEVELOPMENT_TEAM"] = "UF2FQ56Q4Q"
          end
      end
    end
    # End of added lines
  end
end

```

#### `AppDelegate.m`:

```objc
#import "AppDelegate.h"

#import <React/RCTLinkingManager.h>
#import <Firebase.h>
#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>
#import "RNSplashScreen.h" // here

#import "AppName-Swift.h" // here, change project name to yours

#import <React/RCTAppSetupUtils.h>

#if RCT_NEW_ARCH_ENABLED
#import <React/CoreModulesPlugins.h>
#import <React/RCTCxxBridgeDelegate.h>
#import <React/RCTFabricSurfaceHostingProxyRootView.h>
#import <React/RCTSurfacePresenter.h>
#import <React/RCTSurfacePresenterBridgeAdapter.h>
#import <ReactCommon/RCTTurboModuleManager.h>

#import <react/config/ReactNativeConfig.h>

@interface AppDelegate () <RCTCxxBridgeDelegate, RCTTurboModuleManagerDelegate> {
  RCTTurboModuleManager *_turboModuleManager;
  RCTSurfacePresenterBridgeAdapter *_bridgeAdapter;
  std::shared_ptr<const facebook::react::ReactNativeConfig> _reactNativeConfig;
  facebook::react::ContextContainer::Shared _contextContainer;
}
@end
#endif

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  RCTAppSetupPrepareApp(application);

//  rnfirebase
  [FIRApp configure];
  
  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];

#if RCT_NEW_ARCH_ENABLED
  _contextContainer = std::make_shared<facebook::react::ContextContainer const>();
  _reactNativeConfig = std::make_shared<facebook::react::EmptyReactNativeConfig const>();
  _contextContainer->insert("ReactNativeConfig", _reactNativeConfig);
  _bridgeAdapter = [[RCTSurfacePresenterBridgeAdapter alloc] initWithBridge:bridge contextContainer:_contextContainer];
  bridge.surfacePresenter = _bridgeAdapter.surfacePresenter;
#endif

  UIView *rootView = RCTAppSetupDefaultRootView(bridge, @"AppName", nil);

  if (@available(iOS 13.0, *)) {
    rootView.backgroundColor = [UIColor systemBackgroundColor];
  } else {
    rootView.backgroundColor = [UIColor whiteColor];
  }

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];

  //set rootView background color
  rootView.backgroundColor = [UIColor colorWithRed: 1.00 green: 0.94 blue: 0.94 alpha: 1.00]; // change backgroundColor

  Dynamic *t = [Dynamic new];
  UIView *animationUIView = (UIView *)[t createAnimationViewWithRootView:rootView lottieName:@"logo_animated"]; // change lottieName to your lottie files name

  //instead of rootView, create a full screen UIView called animationContainerView with the same background color as the rootView
  UIView *animationContainerView = [[UIView alloc] initWithFrame:rootView.frame];
  animationContainerView.backgroundColor = [UIColor colorWithRed: 1.00 green: 0.94 blue: 0.94 alpha: 1.00]; // change backgroundColor

  //add animationUIView to animationContainerView
  [animationContainerView addSubview:animationUIView];

  //register animationContainerView to RNSplashScreen
  [RNSplashScreen showLottieSplash:animationContainerView inRootView:rootView];

  // casting UIView type to AnimationView type
  AnimationView *animationView = (AnimationView *) animationUIView;

  // play
  [t playWithAnimationView:animationView];

  // If you want the animation layout to be forced to remove when hide is called, use this code
  [RNSplashScreen setAnimationFinished:true];
  
  return YES;
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index"];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

#if RCT_NEW_ARCH_ENABLED

#pragma mark - RCTCxxBridgeDelegate

- (std::unique_ptr<facebook::react::JSExecutorFactory>)jsExecutorFactoryForBridge:(RCTBridge *)bridge
{
  _turboModuleManager = [[RCTTurboModuleManager alloc] initWithBridge:bridge
                                                             delegate:self
                                                            jsInvoker:bridge.jsCallInvoker];
  return RCTAppSetupDefaultJsExecutorFactory(bridge, _turboModuleManager);
}

#pragma mark RCTTurboModuleManagerDelegate

- (Class)getModuleClassFromName:(const char *)name
{
  return RCTCoreModulesClassProvider(name);
}

- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:(const std::string &)name
                                                      jsInvoker:(std::shared_ptr<facebook::react::CallInvoker>)jsInvoker
{
  return nullptr;
}

- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:(const std::string &)name
                                                     initParams:
                                                         (const facebook::react::ObjCTurboModule::InitParams &)params
{
  return nullptr;
}

- (id<RCTTurboModule>)getModuleInstanceFromClass:(Class)moduleClass
{
  return RCTAppSetupDefaultModuleFromClass(moduleClass);
}

#endif

- (BOOL)application:(UIApplication *)application
   openURL:(NSURL *)url
   options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
{
  return [RCTLinkingManager application:application openURL:url options:options];
}

@end

```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
System:
    OS: macOS 13.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 91.50 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 19.5.0 - /opt/homebrew/bin/node
    Yarn: 1.22.19 - ~/node_modules/.bin/yarn
    npm: 9.3.1 - /opt/homebrew/bin/npm
    Watchman: Not Found
  Managers:
    CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 22.2, iOS 16.2, macOS 13.1, tvOS 16.1, watchOS 9.1
    Android SDK:
      API Levels: 30, 31, 33
      Build Tools: 30.0.2, 30.0.3, 31.0.0, 33.0.0, 33.0.1
      System Images: android-21 | Google APIs ARM 64 v8a, android-23 | Google APIs ARM 64 v8a, android-28 | Google ARM64-V8a Play ARM 64 v8a, android-29 | Google APIs ARM 64 v8a, android-29 | Google Play ARM 64 v8a, android-30 | Google Play ARM 64 v8a, android-31 | Google APIs ARM 64 v8a, android-31 | Google Play ARM 64 v8a, android-33 | Google APIs ARM 64 v8a
      Android NDK: Not Found
  IDEs:
    Android Studio: 2022.1 AI-221.6008.13.2211.9477386
    Xcode: 14.2/14C18 - /usr/bin/xcodebuild
  Languages:
    Java: 18.0.1 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2 
    react-native: 0.68.6 => 0.68.6 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [x] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `17.0.0`
- **`Firebase` module(s) you're using that has the issue:**
  - `Messaging`
- **Are you using `TypeScript`?**
  - `N` & `N/A`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

### PR DESCRIPTION
**Pull request by tomwanzek**
### Description

This PR addresses cyclical import dependencies introduced in the modularization of RN Firebase Analytics and Messaging.

### Related issues

Fixes #6897

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [ ] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

🔥

### CODE REVIEW

**Comment by tomwanzek** [comment]
See comment https://github.com/invertase/react-native-firebase/issues/6897#issuecomment-1445423440 with some initial questions to make sure my PR meets the mark.

**Comment by tomwanzek** [comment]
Any update on reviewing this?

**Comment by mikehardy** [comment]
Sorry for the delay! Require cycles have been an issue for quite some time, thanks a  bunch for taking it up

I'm going to paste in your questions from the related issue as they should be in the PR discussion I think?

> I'm not familiar with the modularization approach/next steps RN Firebase is undergoing right now to know if there is any architectural reason, why the creation of the Firebase root instance constant cannot be moved into the /modular/index.js file. I.e. ensuring that all imports are always directed from /modular to /lib. Or if an alternative approach would be needed were the root instance creation and export should be done in a separate file, possibly co-located with the lib/index.js file and both /modular/index.js and /lib/index.js import from that file to break the cycle.

I don't believe there is any specific reason... modular will be the future so having it be the location of record seems fine to me

> In both the Analytics and Messaging current implementation, there are comments left in the code. I'm not cure if that means there was still work to be done which affects the proposed fix, or if those are obsolete legacy comments which can be removed. See:
> https://github.com/invertase/react-native-firebase/blob/main/packages/analytics/lib/index.js#L785-L788
> https://github.com/invertase/react-native-firebase/blob/main/packages/messaging/lib/index.js#L541-L544

Those look like remnants to me - seems okay to delete. Always git history if needed for resurrection

> Not directly related, I noticed that the Podfile.lock versions of the Firebase related pods are not current. I.e. running yarn tests:ios:pod:install will create a diff between versions 17.1.0 to 17.3.1. Is there any reason, why that diff was not previously committed? I discarded it in my draft PR for the time being 🤔

Those only get an update after the release, when I go in and commit the results of a `yarn tests:ios:pod:install` which I tend to do quickly as a habit but sometimes there is a lag.

Either way, ideally your PR had just a single commit that broke the require cycle, no need to merge from HEAD or update the podfile lock

**Review by mikehardy — APPROVED** [review]
Seems like a reasonable approach - per your questions it appears deleting the comments might be worthwhile but I'll leave that as implementors choice. The actual code in here seems like it should work fine

**Comment by mikehardy** [comment]
Please accept my continued apologies - I am still a bit backlogged but hope to give this a look through + merge shortly

**Comment by tomwanzek** [comment]
@mikehardy No worries, I ran into the same bottleneck over the past couple of weeks and did not have a chance to follow up.

One question about:
> it appears deleting the comments might be worthwhile but I'll leave that as implementors choice

Would you like me to push a commit then to remove the comments? Not sure, if by "implementator" you referred to me or the original implementer who added the comments 😄 

Also, it looks like the failed `Documentation / Spelling & Grammar (pull_request)` check is unrelated to the changes in this PR. I can sync the branch against latest master, which may address that one implicitly, if that was fixed elsewhere.

**Comment by tomwanzek** [comment]
Synced my fork to upstream `main` and updated this branch.

**Comment by tomwanzek** [comment]
The main issue that I see right now, is that the same flawed import pattern which this PR tries to fix is/keeps getting applied in other Firebase Packages which are migrated to "modular".

Now that I merged in synced `main`, I see imports of `import { firebase } from '..';` with likely the same circular dependency impact or at least an anti-pattern in:
* `app-check`
* `functions`
* `perf`
* `remote-config`
* `storage`

Possibily it is also in some working branches which have not been merged for other packages.

@russellwheatley @Lyokone I'm adding you to this thread/PR as I found commits with your handles on related lines. Not sure, if anyone else is currently working on this matter who should be added for awareness.

In short, IMHO we should rectify the circular dependency pattern in the existing places of the code base and stop using it going forward, if we are all in agreement that the pattern proposed in this PR is preferrable.

Please, advice at your earliest. I can update this PR to address the aforementioned 5 packages in addition to `messaging` and `analytics`.

**Comment by tomwanzek** [comment]
@mikehardy Hope you're doing well. I'm wondering, if there is any update to the question, I was wondering about in https://github.com/invertase/react-native-firebase/pull/6940#issuecomment-1493403450 . @russellwheatley @Lyokone , do you have a particular take on how to proceed so we have a consistent solution across all packages?

Both the PR and the issue got stale marked, so I would hope we can resuscitate  the matter and close it out in a suitable way. Thanks in advance for giving it a thought and letting me know. 🙇

**Comment by tomwanzek** [comment]
Closing this PR due to inability to receive additional guidance and the related issue having been repeatedly marked stale. Feel free to re-open it, if it is still desirable.

**Comment by mikehardy** [comment]
Really sorry @tomwanzek I would like to keep this on the radar but haven't had sufficient time
